docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,210 @@
+
+## Goals
+
+* Make it more clear about the main workflow of the series deletion process.
+* Perform the actual series deletion using compactor in a scalable way.
+
+## Non-Goals
+
+* Allow setting different retentions per series. This feature can be built on top of this work ideally but it is out of scope of this proposal.
+
+## How
+
+The main workflow of the series deletion will look like this:
+
+1. User create a tombstone file via a CLI tool and the tombstone file is uploaded to the object storage.
+2. Store Gateway syncs tombstone files periodically and makes the block series based on the tombstones.
+3. Tombstone files have a pending period, which can be either specified by each tombstone or configured globally. Before the pending period is over, the tombstone file will remain in the object storage and user can remove a tombstone file using the CLI tool. Once the pending period is over, the compactor starts processing the tombstones and performs deletion.
+4. For each tombstone file, once all the matched blocks are processed, compactor will delete the tombstone file from the object storage.
+
+### Tombstone File
+
+The design of the tombstone file mainly follows the previous proposal. Tombstone file is immutable. It is using a custom JSON format, single file per request and global, which means it applies to all the blocks. For the detailed considerations and tradeoffs, please refer to the original proposal.
+
+The file format is as follows:
+``` bash
+cat 01G61GC0Q957TVZ2FC8S6Y2ZK7.json
+{
+        "ulid": "01G61GC0Q957TVZ2FC8S6Y2ZK7",
+        "matchers": [
+                "__name__=\"k8s_app_metric39\"",
+                "cluster=\"us\"",
+                "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""
+        ],
+        "minTime": -62167219200000,
+        "maxTime": 253402300799000,
+        "author": "Ben Ye",
+        "reason": "not specified"
+}
+```
+
+Similar to the block `meta.json` file, the tombstone file is named using ULID which is basically its creation time. For other fields we have:
+* **label matchers**: label matchers to match the block and series. Like this example, blocks can be matched using external labels and blocks can be matched using a special `__block_id` label. All other labels are applied to series. This design makes it easy to support multi-tenant environment and match specific blocks.
+* **start timestamp**
+* **end timestamp**: end timestamp should be the tombstone creation time if not specified, otherwise the compactor needs to wait for a long time to make sure all matched blocks are processed. Considering the tombstone pending period, the end timestamp can be set to `(now + pending period)` as well. We don't consider history blocks backfilling now.
+* **author**
+* **reason**
+
+### Store Gateway Masks Series based on Tombstones
```

Comment:
Yes, it was implemented in https://github.com/thanos-io/thanos/pull/3075 but hasn't been finished. AFAIU, we agree on the high level idea of the initial proposal but for the implementation details, since the pr is not merged I won't consider it as implemented.